src/analysis/retail/druid/balance/modules/spells/GatheringStarstuff.tsx

```diff
@@ -0,0 +1,263 @@
+    this.listCastStarfire = [];
+    this.listCastStarfireTimestamps = [];
+
+    Object.values(GATHERING_STARSTUFF.AFFECTED_CAST).forEach((spell) => {
+      this.gastBuffedAbilities[spell.id] = 0;
+    });
+
+    this.addEventListener(
+      Events.cast.by(SELECTED_PLAYER).spell(GATHERING_STARSTUFF.AFFECTED_CAST),
+      this.onCast,
+    );
+    this.addEventListener(
+      Events.damage.by(SELECTED_PLAYER).spell(GATHERING_STARSTUFF.AFFECTED_DAMAGE),
+      this.onDamage,
+    );
+    this.addEventListener(
+      Events.removebuff.by(SELECTED_PLAYER).spell(SPELLS.CONVOKE_SPIRITS),
+      this.onEndConvoke,
+    );
+
+    this.addEventListener(
+      Events.applybuff.by(SELECTED_PLAYER).spell(SPELLS.GATHERING_STARSTUFF),
+      this.handleStacks,
+    );
+    this.addEventListener(
+      Events.applybuffstack.by(SELECTED_PLAYER).spell(SPELLS.GATHERING_STARSTUFF),
+      this.handleStacks,
+    );
+    this.addEventListener(
+      Events.removebuff.by(SELECTED_PLAYER).spell(SPELLS.GATHERING_STARSTUFF),
+      this.handleStacks,
+    );
+    this.addEventListener(
+      Events.refreshbuff.by(SELECTED_PLAYER).spell(SPELLS.GATHERING_STARSTUFF),
+      this.handleStacks,
+    );
+    this.addEventListener(Events.fightend, this.handleStacks);
+  }
+
+  // As stacks are consumed when the cast end but the damage are done when the spell lands, we need
+  // to match CastEvent (on which we obtain the number of stacks) with their respective DamageEvent.
+  // To do so we list all CastEvent and when a DamageEvent is read we check with the first CastEvent
+  // on the list if they are in an acceptable time frame. If so we know how many stacks are associated
+  // with a DamageEvent. We can then discard the first CastEvent and go on with our Event. If not,
+  // we either discard the damage event if it's to early (can happen with precast or convoke
+  // cast) or the cast event if it's to late (can happen with a targt dying after the cast but
+  // before the spell lands).
+
+  onCast(event: CastEvent) {
+    const stacks = this.selectedCombatant.getBuff(SPELLS.GATHERING_STARSTUFF.id)?.stacks;
+    this.gastCurrentStacks = stacks ?? 0;
+    if (this.gastCurrentStacks > 0) {
+      //Track which abilities are buffed
+      this.gastBuffedAbilities[event.ability.guid] += 1;
+    }
+    if (event.ability.guid === SPELLS.WRATH_MOONKIN.id) {
+      //Append a Wrath CastEvent and keep its timestamp
+      this.listCastWrath.push(this.gastCurrentStacks);
+      this.listCastWrathTimestamps.push(event.timestamp);
+    }
+    if (event.ability.guid === SPELLS.STARFIRE.id) {
+      //Append a Starfire CastEvent and keep its timestamp
+      this.listCastStarfire.push(this.gastCurrentStacks);
+      this.listCastStarfireTimestamps.push(event.timestamp);
+    }
+  }
+
+  onDamage(event: DamageEvent) {
+    let checkCastList = true;
+
+    //Wrath
+    if (event.ability.guid === SPELLS.WRATH_MOONKIN.id) {
+      while (checkCastList) {
```

Comment:
This is setting up a potential infinite loop with even the slighted logic bug. Can we refactor this so it's not using a while loop?